File: stepik-union/src/main/java/org/stepik/plugin/actions/navigation/StudyStepNavigationAction.java
package org.stepik.plugin.actions.navigation;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.actions.StudyActionWithShortcut;
import org.stepik.core.courseFormat.StudyNode;

import javax.swing.*;


abstract class StudyStepNavigationAction extends StudyActionWithShortcut {
    StudyStepNavigationAction(@Nullable String text, @Nullable String description, @Nullable Icon icon) {
        super(text, description, icon);
    }

    protected abstract void navigateStep(@NotNull final Project project);

    @Override
    public void actionPerformed(AnActionEvent e) {
        Project project = e.getProject();
        if (project == null) {
            return;
        }
        navigateStep(project);
    }

    protected abstract StudyNode getTargetStep(@Nullable final StudyNode sourceStepNode);

    @Override
    public void update(AnActionEvent e) {
        final Presentation presentation = e.getPresentation();
        presentation.setEnabled(false);

        Project project = e.getProject();
        if (!StepikProjectManager.isStepikProject(project)) {
            return;
        }

        StudyNode selected = StepikProjectManager.getSelected(project);
        StudyNode target = getTargetStep(selected);
        boolean enabled = selected == null || target != null;

        if (StepikProjectManager.isAdaptive(project)) {
            enabled = enabled && selected != null && target.getParent() == selected.getParent();
        }
        presentation.setEnabled(enabled);
    }
}

File: stepik-union/src/main/java/org/stepik/core/actions/navigation/StudyStepNavigationAction.kt
package org.stepik.core.actions.navigation

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import org.stepik.core.StepikProjectManager
import org.stepik.core.actions.StudyActionWithShortcut
import org.stepik.core.courseFormat.StudyNode
import javax.swing.Icon


abstract class StudyStepNavigationAction(text: String?,
                                         description: String?,
                                         icon: Icon?) :
        StudyActionWithShortcut(text, description, icon) {

    protected abstract fun navigateStep(project: Project)

    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project ?: return
        navigateStep(project)
    }

    protected abstract fun getTargetStep(sourceStepNode: StudyNode<*, *>?): StudyNode<*, *>?

    override fun update(e: AnActionEvent?) {
        val presentation = e!!.presentation
        presentation.isEnabled = false

        val project = e.project!!
        if (!StepikProjectManager.isStepikProject(project)) {
            return
        }

        val selected = StepikProjectManager.getSelected(project)
        val target = getTargetStep(selected)
        var enabled = selected == null || target != null

        if (StepikProjectManager.isAdaptive(project)) {
            enabled = enabled && selected != null && target!!.parent === selected.parent
        }
        presentation.isEnabled = enabled
    }
}
